statefun/fun.py

```diff
@@ -1,77 +1,190 @@
-
-    # send it to the exit topic
-    response = compute_response(name= request.name)
-    egress_message = kafka_egress_record(topic='greetings', key=request.name, value=response)
-    context.pack_and_send_egress("example/greets", egress_message)
-    print(f"Sent message {response} to topic greetings", flush=True)
-
-def compute_response(name):
-    response = ExampleResponse()
-    response.message = "Hello! This is a response for user {}".format(name)
-    return response
-
-@functions.bind("example/request")
-def send_request(context, msg : typing.Union[ExampleRequest, ExampleResponse]):
-    if isinstance(msg, ExampleRequest):
-        print("Got a new request!, Sending message to other function", flush=True)
-        m = ExampleRequest()
-        m.message = msg.message
-        context.pack_and_send("example/reply", "id", m)
-    
-    elif isinstance(msg, ExampleResponse):
-        print(f"Got an example response with message {msg.message}", flush =True)
-
-@functions.bind("example/reply")
-def send_reponse(context, msg: ExampleRequest):
-    print(f"Got a message: {msg.message} replying...", flush =True)
-    m = ExampleResponse()
-    m.message = "repliying to you from example/reply"
-    context.pack_and_reply(m)
-
-
-# Use the handler and expose the endpoint
-handler = RequestReplyHandler(functions)
-
-from flask import Flask, request, jsonify, make_response
-
-app = Flask(__name__)
-
-@app.route('/statefun', methods=['POST'])
-def handle():
-    response_data = handler(request.data)
-    response = make_response(response_data)
-    response.headers.set('Content-Type', 'application/octet-stream')
-    return response
-
-@app.route('/')
-def welcome():
-    return "Hello world!"
-
-
-if __name__ == "__main__":
-    app.run()
-
-
+""" File including the functions served by the endpoint """
+import typing
+import logging
+import json
+
+from users_pb2 import CreateUserRequest
+from users_pb2 import UserRequest, UserResponse
+
+# to be used internally as state or message from 
+# one function to the other
+from users_pb2 import UserData, Count, CreateUserWithId
+
+from statefun import StatefulFunctions
+from statefun import RequestReplyHandler
+from statefun import kafka_egress_record
+
+USER_EVENTS_TOPIC = "user-events"
+
+functions = StatefulFunctions()
+
+FORMAT = '[%(asctime)s] %(levelname)-8s %(message)s'
+logging.basicConfig(level=logging.DEBUG, format=FORMAT)
+
+logger = logging.getLogger()
+
+
+# Functions to deal with user management
+
+# Function to create users
+# extracts the next free user id from its state and 
+# sends a creation request to that new user function
+@functions.bind("users/create")
+def create_user(context, create_user_request : CreateUserRequest):
+    """ Creates a user by sending a message to the user function 
+    - Only has one state (int) that saves the current id to be 
+    asigned to the next user """
+
+    logger.debug("Creating user...")

```

Comment:
We need to check if these `debug` calls are evaluated at runtime (even though debug is disabled). I know for Java/Scala they do some trick that if debug is disabled, that these lines are somehow removed/skipped at runtime. I'm not sure how that works for Python. Worst-case it evaluates all the debug lines and then has some internal check if debug mode is enabled.. maybe that is a bit of an expensive operation if we want these stateful functions to be performant.

Reply:
Mmmm indeed, hadn't thought about that, we can take those out for sure